### ISSUE TITLE
Automatically resolve Mach-O Segments

### ISSUE THREAD

**Issue by duraki** [NationalSecurityAgency/ghidra#1823]
**This is proposal for automatic segment fixes on Mach-O files.**

I'm running side by side IDA Free and Ghidra SRE. My target is a Mach-O file:

```
$ file target
target: Mach-O 64-bit executable x86_64
```

**Describe the problem:**
* Loading the app in **Ghidra** results in *obfuscated code* listing (ie. too many idbs).
* Loading the app in **IDA Free** results in *deobfuscated code*.

**What is happening:**
The thing with the target application is that section file offset is invalid, and it's placed as a way to hinder/harder the analysis. Therefore loading the target app. in Ghidra SRE does not deobfuscate the file while analysing.  

IDA, as a difference, detects this prior to analysis, and therefore automatically fixes the sections to reflect proper entrypoints.

```
Loading file 'target' into database...
Detected file format: Mach-O file (EXECUTE). X86_64
...
Autoanalysis subsystem has been initialized.
Type library 'macosx64' loaded. Applying types...
Types applied to 0 names.
0x10000A6F0: section file offset invalid: (5242; expected A6F0); possibly the file was modified to hinder analysis!
  0. Creating a new segment  (000000010000A6F0-000000010044A776) ... ... OK
0x10044A776: section file offset invalid: (4460BF; expected 44A776); possibly the file was modified to hinder analysis!
  1. Creating a new segment  (000000010044A776-000000010044B5CE) ... ... OK
0x10044B5D0: section file offset invalid: (44ADC0; expected 44B5D0); possibly the file was modified to hinder analysis!
  2. Creating a new segment  (000000010044B5D0-000000010044CDB4) ... ... OK
0x10044CDB4: section file offset invalid: (44C84B; expected 44CDB4); possibly the file was modified to hinder analysis!

... etc until all sections are resolved ...
```

**Describe the solution you'd like:**
As of now, one can use **MachOView** to manually fix segments and reload+analyse the app in Ghidra. I'm not aware of any Ghidra option that can do that manually (if so, I'd like to know).

As a solution, therefore I'd like to have something that IDA offers natively. 

**Describe alternatives you've considered**
I'm not sure there is Ghidra native way to manually fix the sections prior to analysis. Anyway, MachOView is an alternative for now.

**Additional context**
*None*

Thank you.

**Comment by ryanmkurtz** [comment]
Can you provide the binary?

**Comment by duraki** [comment]
* [Official Installation (DMG)](https://d2ap6ypl1xbe4k.cloudfront.net/Hopper-4.5.24-demo.dmg)
* [Binary only](https://github.com/duraki/hd-binary/blob/master/Hopper%20Disassembler%20v4)

(The binary is rather large; 25MB). It's a commercial product but it's allowed to be reverse engineered, legally, as per their ToS.

**Comment by ryanmkurtz** [comment]
Can you provide the binary?

**Comment by duraki** [comment]
Yes, absolutely. Sorry for the delay.

Here is a binary of Hopper Disassembler Demo version. 

`https://wormhole.app/bldxP#ttgvWmVopRBlaSIghx6JgA`

(The file will be deleted in 24h, please use [official link](https://www.hopperapp.com/download) as an alternative.

```
$ md5 /Applications/Hopper\ Disassembler\ Demo.app/Contents/MacOS/Hopper\ Disassembler\ v4
MD5 (/Applications/Hopper Disassembler Demo.app/Contents/MacOS/Hopper Disassembler v4) = 234d38d91d4f24bbff7830f41472777f
```

**Ghidra Function Export** (incorrect signatures)
![Screenshot 2022-05-23 at 16 24 06](https://user-images.githubusercontent.com/5892610/169841336-13817a54-7bb0-4cb9-96a9-7b6fdef904b2.png)

**Hopper Functions and Labels** (correct signatures + demangle)
![Screenshot 2022-05-23 at 16 27 08](https://user-images.githubusercontent.com/5892610/169841944-fa430033-4b4d-48a3-88b8-5af5f2e79e9e.png)

**Comment by duraki** [comment]
~ slide 29 as a ref// https://iokit.racing/machotricks.pdf

**Comment by ryanmkurtz** [comment]
Thanks.  Is this the same exact binary you referenced in your original post (do the section file offsets in the output still apply)?

**Comment by ryanmkurtz** [comment]
Also, do you know if there is a tool that exists to do this obfuscation?

**Comment by duraki** [comment]
> Thanks. Is this the same exact binary you referenced in your original post (do the section file offsets in the output still apply)?

Not really, no. The original post was older version of Hopper, not sure which one exactly - but the proof/evidence stays the same when it comes to Ghidra functionality. I can give you a new offsets per IDA console output, just need a bit time to find someone who have IDA by hand.

> Also, do you know if there is a tool that exists to do this obfuscation?

Good question. I will let you know, but as far I'm aware there isn't any

**Comment by duraki** [comment]
No worries, glad I can help. That is true, as per Mach-O they aren't strict, so some variability is possible. I think you can use dyld's dsc extractor sample to pre-reference those (albeit only on MacOS) 

ref// https://opensource.apple.com/source/dyld/dyld-750.5/launch-cache/dsc_extractor.cpp.auto.html
